src/integration-test/java/com/commercetools/sync/integration/sdk2/services/impl/CartDiscountServiceImplIT.java

```diff
@@ -0,0 +1,413 @@
+    assertThat(errorCallBackMessages).isEmpty();
+  }
+
+  @Test
+  void fetchMatchingCartDiscountsByKeys_WithBadGateWayExceptionAlways_ShouldFail() {
+    // Mock sphere client to return BadGatewayException on any request.
+    final ProjectApiRoot spyClient = spy(CTP_TARGET_CLIENT);
+    final ByProjectKeyCartDiscountsRequestBuilder byProjectKeyCartDiscountsRequestBuilder = mock();
+    when(spyClient.cartDiscounts()).thenReturn(byProjectKeyCartDiscountsRequestBuilder);
+    final ByProjectKeyCartDiscountsGet byProjectKeyCartDiscountsGet = mock();
+    when(byProjectKeyCartDiscountsRequestBuilder.get()).thenReturn(byProjectKeyCartDiscountsGet);
+    when(byProjectKeyCartDiscountsGet.withWhere(anyString()))
+        .thenReturn(byProjectKeyCartDiscountsGet);
+    when(byProjectKeyCartDiscountsGet.withPredicateVar(anyString(), any()))
+        .thenReturn(byProjectKeyCartDiscountsGet);
+    when(byProjectKeyCartDiscountsGet.withWithTotal(anyBoolean()))
+        .thenReturn(byProjectKeyCartDiscountsGet);
+    when(byProjectKeyCartDiscountsGet.withLimit(anyInt())).thenReturn(byProjectKeyCartDiscountsGet);
+    when(byProjectKeyCartDiscountsGet.execute())
+        .thenReturn(
+            CompletableFutureUtils.exceptionallyCompletedFuture(
+                new BadGatewayException(500, "", null, "", null)))
+        .thenCallRealMethod();
+
+    final CartDiscountSyncOptions spyOptions =
+        CartDiscountSyncOptionsBuilder.of(spyClient)
+            .errorCallback(
+                (exception, oldResource, newResource, actions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                  errorCallBackExceptions.add(exception);
+                })
+            .build();
+
+    final CartDiscountService spyCartDiscountService = new CartDiscountServiceImpl(spyOptions);
+
+    final Set<String> cartDiscountKeys = new HashSet<>();
+    cartDiscountKeys.add(CART_DISCOUNT_KEY_1);
+
+    // test and assert
+    assertThat(errorCallBackExceptions).isEmpty();
+    assertThat(errorCallBackMessages).isEmpty();
+    assertThat(spyCartDiscountService.fetchMatchingCartDiscountsByKeys(cartDiscountKeys))
+        .failsWithin(10, TimeUnit.SECONDS)
+        .withThrowableOfType(ExecutionException.class)
+        .withCauseExactlyInstanceOf(BadGatewayException.class);
+  }
+
+  @Test
+  void createCartDiscount_WithValidCartDiscount_ShouldCreateCartDiscount() {
+    // preparation
+    final CartDiscountDraft newCartDiscountDraft =
+        CartDiscountDraftBuilder.of()
+            .name(CART_DISCOUNT_NAME_2)
+            .cartPredicate(CART_DISCOUNT_CART_PREDICATE_2)
+            .value(CART_DISCOUNT_VALUE_DRAFT_2)
+            .target(CART_DISCOUNT_TARGET_2)
+            .sortOrder(SORT_ORDER_2)
+            .requiresDiscountCode(false)
+            .key(CART_DISCOUNT_KEY_2)
+            .isActive(false)
+            .build();
+
+    final ProjectApiRoot spyClient = spy(CTP_TARGET_CLIENT);
```

Comment:
why do you need a `spy` here?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/1027/commits/c4b802dbfa39528e0314fe18b21af832e8b0e88c